akka-http/src/main/scala/akka/http/scaladsl/server/JsonEntityStreaming.scala

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.http.scaladsl.server
+
+import akka.NotUsed
+import akka.actor.ActorSystem
+import akka.http.impl.util.JavaMapping
+import akka.http.scaladsl.model.{ ContentType, ContentTypes }
+import akka.http.scaladsl.server.directives.FramedEntityStreamingDirectives.SourceRenderingMode
+import akka.stream.scaladsl.{ Flow, Framing }
+import akka.util.ByteString
+import com.typesafe.config.Config
+
+import scala.collection.immutable
+
+/**
+ * Same as [[akka.stream.scaladsl.Framing]] but additionally can express which [[ContentType]] it supports,
+ * which can be used to reject routes if content type does not match used framing.
+ */
+abstract class FramingWithContentType extends Framing {
```

Comment:
While I'm not in love with this idea, it allows us to have nice "eager" rejections if we know we won't be able to apply framing (by being lazy, we check if it works, and then apply).

This relates to the ideas about making Unmarshaller "lazy" (check it it can work, then apply), which is the large change in unmarshalling we should think about before going stable.